graylog2-server/src/main/java/org/graylog2/shared/security/HttpHeadersToken.java

```diff
@@ -0,0 +1,51 @@
+/**
+ * This file is part of Graylog.
+ *
+ * Graylog is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Graylog is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.shared.security;
+
+import org.apache.shiro.authc.HostAuthenticationToken;
+
+import javax.ws.rs.core.MultivaluedMap;
+
+public class HttpHeadersToken implements HostAuthenticationToken {
+
+    private final MultivaluedMap<String, String> httpHeaders;
+    private final String host;
+
+    public HttpHeadersToken(MultivaluedMap<String, String> httpHeaders, String host) {
+        this.httpHeaders = httpHeaders;
+        this.host = host;
+    }
+
+    @Override
+    public Object getPrincipal() {
+        return httpHeaders.get("remote_user");
+    }
+
+    @Override
+    public Object getCredentials() {
+        return httpHeaders.get("remote_credentials");
```

Comment:
These are really only return values because the interface requires it.
The intention for a HttpHeadersToken is that a plugin (such as the SSO one) can pick any header(s) it likes to get the relevant information. The only user of this token is not even using `getPrincipal`/`getCredentials`.
However, I think this should be clearer, so I'll add documentation comments to that effect.
I think `remote_user` is a good choice as it is often used, `remote_credentials` is probably obscure. I'll check if we can simply return null here safely.